PresentData/PresentMonTraceConsumer.cpp

```diff
@@ -316,11 +335,17 @@ void PMTraceConsumer::HandleDxgkQueueSubmit(
         TRACK_PRESENT_PATH(eventIter->second);
         DebugModifyPresent(*eventIter->second);
 
+        // TODO: What happens if there already is an item in mPresentsBySubmitSequence?

```

Comment:
Like I said below, submit sequences are unique.

Reply:
Heads up that I hit asserting on this condition (SubtmitSequence key doesn't already exist) when selfhosting PresentMon over night. Unfortunately the debugger connection seems to have dropped somehow, so I don't have any other info.

![image](https://user-images.githubusercontent.com/1754267/98285843-fc461800-1f57-11eb-89a7-74d7d535e3d5.png)

Will continue selfhosting to see if I hit it again.

Reply:
Just hit this again. Not sure if it's related, but I was moving a menu item in Microsoft Todo Windows App when this hit. Does anything stand out?

[Link to dump](https://www.dropbox.com/s/d37w5rlosgo67zt/AssertOnTwoSubmitQueues.zip?dl=0)

Event we want to insert:
```
    [+0x000] QpcTime          : 0x7cac0f2bb8 [Type: unsigned __int64]
    [+0x008] ProcessId        : 0x641c [Type: unsigned int]
    [+0x00c] ThreadId         : 0x1d34 [Type: unsigned int]
    [+0x010] TimeTaken        : 0x0 [Type: unsigned __int64]
    [+0x018] ReadyTime        : 0x0 [Type: unsigned __int64]
    [+0x020] ScreenTime       : 0x0 [Type: unsigned __int64]
    [+0x028] SwapChainAddress : 0x0 [Type: unsigned __int64]
    [+0x030] SyncInterval     : -1 [Type: int]
    [+0x034] PresentFlags     : 0x0 [Type: unsigned int]
    [+0x038] Hwnd             : 0x0 [Type: unsigned __int64]
    [+0x040] TokenPtr         : 0xffffc2853a3a58c0 [Type: unsigned __int64]
    [+0x048] QueueSubmitSequence : 0x7 [Type: unsigned int]
    [+0x04c] Runtime          : Other (2) [Type: Runtime]
    [+0x050] PresentMode      : Composed_Composition_Atlas (7) [Type: PresentMode]
    [+0x054] FinalState       : Unknown (0) [Type: PresentResult]
    [+0x058] DestWidth        : 0x0 [Type: unsigned int]
    [+0x05c] DestHeight       : 0x0 [Type: unsigned int]
    [+0x060] CompositionSurfaceLuid : 0x0 [Type: unsigned __int64]
    [+0x068] SupportsTearing  : false [Type: bool]
    [+0x069] MMIO             : false [Type: bool]
    [+0x06a] SeenDxgkPresent  : false [Type: bool]
    [+0x06b] SeenWin32KEvents : false [Type: bool]
    [+0x06c] DriverBatchThreadId : 0x0 [Type: unsigned int]
    [+0x070] DwmNotified      : false [Type: bool]
    [+0x071] Completed        : false [Type: bool]
    [+0x078] DependentPresents : { size=0x0 } [Type: std::deque<std::shared_ptr<PresentEvent>,std::allocator<std::shared_ptr<PresentEvent> > >]
    [+0x0a0] TrackingIndex    : 0x1541 [Type: unsigned int]
    [+0x0a8] DxgKrnlHContext  : 0x0 [Type: unsigned __int64]
    [+0x0b0] Win32KPresentCount : 0x0 [Type: unsigned __int64]
    [+0x0b8] Win32KBindId     : 0x0 [Type: unsigned __int64]
    [+0x0c0] LegacyBlitTokenData : 0x0 [Type: unsigned __int64]
    [+0x0c8] PresentInDwmWaitingStruct : false [Type: bool]

```

mPresentsBySubmitSequence:
```
    [<Raw View>]     [Type: std::map<unsigned int,std::shared_ptr<PresentEvent>,std::less<unsigned int>,std::allocator<std::pair<unsigned int const ,std::shared_ptr<PresentEvent> > > >]
    [comparator]     : less [Type: std::_Compressed_pair<std::less<unsigned int>,std::_Compressed_pair<std::allocator<std::_Tree_node<std::pair<unsigned int const ,std::shared_ptr<PresentEvent> >,void *> >,std::_Tree_val<std::_Tree_simple_types<std::pair<unsigned int const ,std::shared_ptr<PresentEvent> > > >,1>,1>]
    [allocator]      : allocator [Type: std::_Compressed_pair<std::allocator<std::_Tree_node<std::pair<unsigned int const ,std::shared_ptr<PresentEvent> >,void *> >,std::_Tree_val<std::_Tree_simple_types<std::pair<unsigned int const ,std::shared_ptr<PresentEvent> > > >,1>]
    [0x0]            : 0x7, {...} [Type: std::pair<unsigned int const ,std::shared_ptr<PresentEvent> >]
    [0x1]            : 0x59a, {...} [Type: std::pair<unsigned int const ,std::shared_ptr<PresentEvent> >]
    [0x2]            : 0x59d, {...} [Type: std::pair<unsigned int const ,std::shared_ptr<PresentEvent> >]
    [0x3]            : 0xbfcbc, {...} [Type: std::pair<unsigned int const ,std::shared_ptr<PresentEvent> >]
    [0x4]            : 0xbfcc1, {...} [Type: std::pair<unsigned int const ,std::shared_ptr<PresentEvent> >]
    [0x5]            : 0xbfcc6, {...} [Type: std::pair<unsigned int const ,std::shared_ptr<PresentEvent> >]
    [0x6]            : 0x6eae1c, {...} [Type: std::pair<unsigned int const ,std::shared_ptr<PresentEvent> >]
    [0x7]            : 0x6eae20, {...} [Type: std::pair<unsigned int const ,std::shared_ptr<PresentEvent> >]
```

Existing event: 
```
    [<Raw View>]     [Type: std::shared_ptr<PresentEvent>]
    [ptr]            : 0x1e140f88c60 [Type: PresentEvent *]
    [control block]  [Type: std::_Ref_count_obj2<PresentEvent> (derived from std::_Ref_count_base)]
    [+0x000] QpcTime          : 0x7cac0ddcbc [Type: unsigned __int64]
    [+0x008] ProcessId        : 0x55dc [Type: unsigned int]
    [+0x00c] ThreadId         : 0x222c [Type: unsigned int]
    [+0x010] TimeTaken        : 0x0 [Type: unsigned __int64]
    [+0x018] ReadyTime        : 0x7cac0de29c [Type: unsigned __int64]
    [+0x020] ScreenTime       : 0x0 [Type: unsigned __int64]
    [+0x028] SwapChainAddress : 0x0 [Type: unsigned __int64]
    [+0x030] SyncInterval     : -1 [Type: int]
    [+0x034] PresentFlags     : 0x0 [Type: unsigned int]
    [+0x038] Hwnd             : 0x0 [Type: unsigned __int64]
    [+0x040] TokenPtr         : 0xffffc2853a3a58c0 [Type: unsigned __int64]
    [+0x048] QueueSubmitSequence : 0x7 [Type: unsigned int]
    [+0x04c] Runtime          : Other (2) [Type: Runtime]
    [+0x050] PresentMode      : Composed_Composition_Atlas (7) [Type: PresentMode]
    [+0x054] FinalState       : Unknown (0) [Type: PresentResult]
    [+0x058] DestWidth        : 0x0 [Type: unsigned int]
    [+0x05c] DestHeight       : 0x0 [Type: unsigned int]
    [+0x060] CompositionSurfaceLuid : 0x0 [Type: unsigned __int64]
    [+0x068] SupportsTearing  : false [Type: bool]
    [+0x069] MMIO             : false [Type: bool]
    [+0x06a] SeenDxgkPresent  : true [Type: bool]
    [+0x06b] SeenWin32KEvents : false [Type: bool]
    [+0x06c] DriverBatchThreadId : 0x0 [Type: unsigned int]
    [+0x070] DwmNotified      : false [Type: bool]
    [+0x071] Completed        : false [Type: bool]
    [+0x078] DependentPresents : { size=0x0 } [Type: std::deque<std::shared_ptr<PresentEvent>,std::allocator<std::shared_ptr<PresentEvent> > >]
    [+0x0a0] TrackingIndex    : 0x153f [Type: unsigned int]
    [+0x0a8] DxgKrnlHContext  : 0x0 [Type: unsigned __int64]
    [+0x0b0] Win32KPresentCount : 0x0 [Type: unsigned __int64]
    [+0x0b8] Win32KBindId     : 0x0 [Type: unsigned __int64]
    [+0x0c0] LegacyBlitTokenData : 0x0 [Type: unsigned __int64]
    [+0x0c8] PresentInDwmWaitingStruct : false [Type: bool]
```

Reply:
Note: event we want to insert came from process 0x641c = Todo.exe,

Preexisting event came from process 0x55dc = DataExchangeHost.exe,

Reply:
It's also interesting that looking at the TrackingIndex, these presents were created one right after another.

Reply:
Huh... looks like the submit sequence is per-context, not global. Unfortunately, some events that use the submit sequence, such as MMIOFlip or the VSync DPCs, don't provide context information, so we can't just simply change the key to (context, submit sequence).

For full determinism, we'd need to split this container in two, one which is keyed on (context, submit sequence) for tracking rendering from submission to completion, and one which is keyed on (adapter, vidpn source/target ID, submit sequence) for tracking flips, I think.

This is a good find. Maybe let's just work around it for now? Though I guess I don't really have a great idea how to work around it.

Reply:
Resolving this one as out of scope for this PR.